### PR TITLE
vSphere Cloud Provider Implementation 

### PR DESCRIPTION
**Pull request by dagnello**
This is the first PR towards implementation for vSphere cloud provider support in Kubernetes (ref. issue #23932).

### CODE REVIEW

**Comment by lavalamp** [comment]
Hi, it's great to see more cloud providers!

Can you take a look at making CLA bot happy with all of the committers?

**Comment by dagnello** [comment]
Hello @lavalamp!

@vipulsabhaya @abithap need Google's CLA signed  (https://cla.developers.google.com/clas)

**Comment by vipulsabhaya** [comment]
I do have a corporate CLA signed.  Maybe an email address issue?

**Comment by lavalamp** [comment]
Looks like @abithap needs to sign a CLA.

**Comment by lavalamp** [comment]
Can you add to this PR a file `pkg/cloudprovider/providers/vsphere/OWNERS` and add appropriate entries? (look at e.g. `pkg/OWNERS` for an example) This is so change PRs can be assigned to you, and so we know who to direct people to if they have questions or file bugs etc.

**Comment by abithap** [comment]
@lavalamp I have signed the CLA. Can you please re-run the CLA test when you get a chance and see if it passes now? thanks

File: pkg/cloudprovider/providers/vsphere/vsphere.go [hunk 1]
```diff
@@ -148,7 +149,15 @@ func (vs *VSphere) LoadBalancer() (cloudprovider.LoadBalancer, bool) {
 
 // Zones returns an implementation of Zones for Google vSphere.
 func (vs *VSphere) Zones() (cloudprovider.Zones, bool) {
-	return nil, true
+	glog.V(1).Info("Claiming to support Zones")
```
Comment (lavalamp) [inline pkg/cloudprovider/providers/vsphere/vsphere.go]: I would omit, bump to V(4), or clarify e.g. add "VSphere:" in front.

File: pkg/cloudprovider/providers/vsphere/vsphere.go [hunk 2]
```diff
@@ -148,7 +149,15 @@ func (vs *VSphere) LoadBalancer() (cloudprovider.LoadBalancer, bool) {
...
+
+	return vs, true
+}
+
+func (vs *VSphere) GetZone() (cloudprovider.Zone, error) {
+	glog.V(1).Infof("Current zone is %v", vs.cfg.Global.Datacenter)
```
Comment (lavalamp) [inline pkg/cloudprovider/providers/vsphere/vsphere.go]: Same

File: pkg/cloudprovider/providers/vsphere/vsphere.go [hunk 3]
```diff
@@ -80,9 +86,56 @@ func init() {
 	})
 }
 
+func readInstanceID(cfg *VSphereConfig) (string, error) {
+	cmd := exec.Command("bash", "-c", `dmidecode -t 1 | grep UUID | tr -d ' ' | cut -f 2 -d ':'`)
+	var out bytes.Buffer
+    cmd.Stdout = &out
```
Comment (lavalamp) [inline pkg/cloudprovider/providers/vsphere/vsphere.go]: nit: spaces vs tabs (presumably fixed in a subsequent commit)

File: pkg/cloudprovider/providers/vsphere/OWNERS [hunk 4]
```diff
@@ -0,0 +1,4 @@
+assignees:
```
Comment (lavalamp) [inline pkg/cloudprovider/providers/vsphere/OWNERS]: Sorry I got this wrong previously-- actually we're asking maintainers to make this line say "maintainers:"

If you can make this change I'll LGTM :)

**Comment by dagnello** [comment]
@lavalamp changes made, let me know if there is something else to update or change

**Comment by lavalamp** [comment]
ok to test

**Comment by lavalamp** [comment]
LGTM. (@dagnello, if you see this before it merges, it would be good to squash the last two commits together.)

**Comment by dagnello** [comment]
@lavalamp updated

**Comment by dagnello** [comment]
@lavalamp trying to update my godep license file (Mac OS X 10.11.4), getting the following error:

hack/update-godep-licenses.sh --create-missing
hack/update-godep-licenses.sh: line 182: declare: -A: invalid option
declare: usage: declare [-afFirtx] [-p] [name[=value] ...]

**Comment by lavalamp** [comment]
@karlkfi Any ideas about the above license error?

**Comment by karlkfi** [comment]
@lavalamp @dagnello That's a bash error. You need bash version 4 or greater to use associative arrays.

**Comment by dagnello** [comment]
@karlkfi thank you, that resolved this issue!  @lavalamp updated the godep licenses file

**Comment by lavalamp** [comment]
@eparis You wrote the godep verifier, right? Can you explain this riddle to us? What is this file? @dagnello says he's using godep v53.

```
Your godeps changes are not reproducible
diff -Naupr '--exclude=symlink' /go/src/k8s.io/kubernetes/Godeps/_workspace/src/github.com/vmware/govmomi/.drone.sec /tmp/gopath.umT3x3/src/k8s.io/kubernetes/Godeps/_workspace/src/github.com/vmware/govmomi/.drone.sec
--- /go/src/k8s.io/kubernetes/Godeps/_workspace/src/github.com/vmware/govmomi/.drone.sec    1969-12-31 16:00:00.000000000 -0800
+++ /tmp/gopath.umT3x3/src/k8s.io/kubernetes/Godeps/_workspace/src/github.com/vmware/govmomi/.drone.sec    2016-05-05 13:37:01.280462336 -0700
@@ -0,0 +1 @@
+eyJhbGciOiJSU0EtT0FFUCIsImVuYyI6IkExMjhHQ00ifQ.FP7BTRxaTbVILW2DBSeJu63eb2Ppa3-gxe5Ja6B5y9iijsqJWkg9k_4iGnojcfgIXz0GVL7XPqdGOFrzQAJWYn0US53t3OINyjwm0J4O0UFHoyLXjwCscfQ3MU_Ha-nVpoWoBll8Omyv4tXykWs2_lCt9LT6h8u7DlYaz2cFns6JFcawRPe-cbN_yuDM7P24wSHNwPGk_Ls1MvR4phmrwCsIhV7e82Be8bVRMU4btbxv1JSqUgGr6vur8JKqM4n0_yKOtBPswcH4_jyzmQABjaKAUfpuznOsfO42t9RyeV7L6IZfFIqgpZMh00DOl3a2I_YdDpydZT1laCGd1_Sn8Q.pCgShcjPktT8Bx48.pbMl3UhYFCe9iVvTCLrxsoOFziShsKSAIzMc4ZphUFb91BvUFATSGOnEYcfFRdCl3Dd5sFzvFcbIiPvDn5klhk-HZMMSHxPbPGutBsSFoPb4WwRPSz9FxybQkXB1DLFpgMGAMfGyjnt1k5420KDyuogwTNUvMOg_vxnH_P7b.zGHv8arI8aMdXX6wyo51RQ
\ No newline at end of file
Removing /tmp/gopath.umT3x3
FAILED   ./hack/../hack/verify-godeps.sh    206s
```

**Comment by karlkfi** [comment]
@lavalamp: The godep verifier makes a tempdir to use as a clean workspace so it can do a godep restore and diff the result against your workspace to make sure it's reproducible.

**Comment by dagnello** [comment]
@lavalamp ok, I haven't made any changes to the Govmomi library, what difference is it seeing to make this check fail?

**Comment by karlkfi** [comment]
Looks like the `Godeps/_workspace/src/github.com/vmware/govmomi/.drone.sec` file contents are different when you do a godep restore. They need to be the same.

**Comment by eparis** [comment]
Its pretty simple. You'll see .drone.sec in https://github.com/vmware/govmomi/blob/master/.drone.sec but that file isn't being included in Godeps/_workspace/src/github.com/vmware/govmomi/

verify-godep is checking is the contents in Godeps are actually the ones that are supposedly referenced and in this case, they are not. that file is missing.

It seems the right solution is to just add that file to `Godeps`

**Comment by dagnello** [comment]
@karlkfi @eparis thank you, looks like that was the difference.  .drone.sec was being skipped by .gitignore by default

**Comment by lavalamp** [comment]
```

=== RUN   TestReadConfig
--- PASS: TestReadConfig (0.00s)
=== RUN   TestNewVSphere
--- SKIP: TestNewVSphere (0.00s)
	vsphere_test.go:87: No config found in environment
=== RUN   TestVSphereLogin
--- SKIP: TestVSphereLogin (0.00s)
	vsphere_test.go:99: No config found in environment
=== RUN   TestZones
--- FAIL: TestZones (0.07s)
	vsphere_test.go:127: Failed to construct/authenticate vSphere: Post https://:/sdk: dial tcp :0: getsockopt: connection refused
=== RUN   TestInstances
--- SKIP: TestInstances (0.00s)
	vsphere_test.go:148: No config found in environment
FAIL
FAIL	k8s.io/kubernetes/pkg/cloudprovider/providers/vsphere	0.253s
```

Looks like a more legit problem this time :)

**Comment by lavalamp** [comment]
LGTM assuming tests will pass :)

**Comment by dagnello** [comment]
@lavalamp build failure details are not accessible, any way I can get that information?

**Comment by lavalamp** [comment]
@pwittrock or @ixdy-- can one of you tell us the correct node e2e status link?

**Comment by lavalamp** [comment]
@k8s-bot node test this issue #IGNORE

**Pull request merged**